ufl/algorithms/apply_coefficient_split.py

```diff
@@ -0,0 +1,202 @@
+"""Apply coefficient split.
+
+This module contains classes and functions to split coefficients defined on mixed function spaces.
+"""
+
+import numpy
+from ufl.classes import Restricted
+from ufl.corealg.map_dag import map_expr_dag
+from ufl.corealg.multifunction import MultiFunction, memoized_handler
+from ufl.domain import extract_unique_domain
+from ufl.classes import (Coefficient, Form, ReferenceGrad, ReferenceValue,
+                         Indexed, MultiIndex, Index, FixedIndex,
+                         ComponentTensor, ListTensor, Zero,
+                         NegativeRestricted, PositiveRestricted,)
+from ufl import indices
+from ufl.checks import is_cellwise_constant
+from ufl.tensors import as_tensor
+
+
+class CoefficientSplitter(MultiFunction):
+
+    def __init__(self, coefficient_split):
+        MultiFunction.__init__(self)
+        self._coefficient_split = coefficient_split
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def modified_terminal(self, o):
+        restriction = None
+        local_derivatives = 0
+        reference_value = False
+        t = o
+        while not t._ufl_is_terminal_:
+            assert t._ufl_is_terminal_modifier_, f"Got {repr(t)}"
+            if isinstance(t, ReferenceValue):
+                assert not reference_value, "Got twice pulled back terminal!"
+                reference_value = True
+                t, = t.ufl_operands
+            elif isinstance(t, ReferenceGrad):
+                local_derivatives += 1
+                t, = t.ufl_operands
+            elif isinstance(t, Restricted):
+                assert restriction is None, "Got twice restricted terminal!"
+                restriction = t._side
+                t, = t.ufl_operands
+            elif t._ufl_terminal_modifiers_:
+                raise ValueError("Missing handler for terminal modifier type %s, object is %s." % (type(t), repr(t)))
```

Comment:
Is there a reason for not using `f-strings` here?

Reply:
No. I will fix it when I tidy this up. Thanks.

Reply:
Fixed.